File: src/main/java/com/asamm/locus/addon/graphhopper/RoutingService.java
package com.asamm.locus.addon.graphhopper;

import android.content.Intent;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

import com.graphhopper.GHRequest;
import com.graphhopper.GHResponse;
import com.graphhopper.GraphHopper;
import com.graphhopper.PathWrapper;
import com.graphhopper.routing.util.EncodingManager;
import com.graphhopper.routing.util.FlagEncoder;
import com.graphhopper.routing.util.FlagEncoderFactory;
import com.graphhopper.routing.weighting.FastestWeighting;
import com.graphhopper.routing.weighting.ShortestWeighting;
import com.graphhopper.routing.weighting.Weighting;
import com.graphhopper.util.Instruction;
import com.graphhopper.util.Parameters;
import com.graphhopper.util.PointList;
import com.graphhopper.util.RoundaboutInstruction;
import com.graphhopper.util.StopWatch;
import com.graphhopper.util.ViaInstruction;
import com.graphhopper.util.shapes.GHPoint;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import locus.api.android.features.computeTrack.ComputeTrackParameters;
import locus.api.android.features.computeTrack.ComputeTrackService;
import locus.api.android.utils.LocusUtils;
import locus.api.objects.enums.PointRteAction;
import locus.api.objects.extra.GeoDataExtra;
import locus.api.objects.extra.Location;
import locus.api.objects.extra.Track;
import locus.api.objects.extra.Waypoint;
import locus.api.utils.Logger;

/**
 * Created by menion on 9. 7. 2014.
 * Class is part of Locus project
 */
public class RoutingService extends ComputeTrackService {

	// tag for logger
    private static final String TAG = "RoutingService";

    // instance of GraphHooper engine
    private GraphHopper mHopper;
	// remember last used map path
	private File mLastRoutingItem;

	// flag if we have "bike2" mode in latest loaded file
	private boolean mBike2Encoding;

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public String getAttribution() {
        return "Powered by <a href=\"http://graphhopper.com/#enterprise\">GraphHopper API</a>";
    }

    @Override
    public int[] getTrackTypes() {
        // initialize graphHopper instance
        if (getGraphHooper() == null) {
            Logger.logW(TAG, "getTrackTypes()," +
                    "unable to initialize GraphHopper instance");
            return null;
        }

        // load encoders
        EncodingManager em = getGraphHooper().getEncodingManager();
		List<Weighting> weightings = getGraphHooper().getCHFactoryDecorator().getWeightings();
        List<Integer> types = new ArrayList<>();

        // get all encoders
        for (FlagEncoder enc : em.fetchEdgeEncoders()) {
            String encType = enc.toString();

            // add car
            switch (encType) {
                case FlagEncoderFactory.CAR:
                	boolean carAdded = false;
                	for (Weighting weighting : weightings) {
						if (weighting instanceof FastestWeighting) {
							addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CAR_FAST, types);
							carAdded = true;
						} else if (weighting instanceof ShortestWeighting) {
							addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CAR_SHORT, types);
							carAdded = true;
						}
					}

					// add basic "car" if no car type exists
					if (!carAdded) {
						addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CAR, types);
					}
                    break;
                case FlagEncoderFactory.MOTORCYCLE:
                    addTrackType(GeoDataExtra.VALUE_RTE_TYPE_MOTORCYCLE, types);
                    break;
                case FlagEncoderFactory.BIKE:
				case FlagEncoderFactory.BIKE2:
                	boolean bikeAdded = false;
					for (Weighting weighting : weightings) {
						if (weighting instanceof FastestWeighting) {
							addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_FAST, types);
							bikeAdded = true;
						} else if (weighting instanceof ShortestWeighting) {
							addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_SHORT, types);
							bikeAdded = true;
						}
					}

					// add basic "bike" if no bike type exists
					if (!bikeAdded) {
						addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE, types);
					}

					// enable "Bike2" encoding
					if (encType.equalsIgnoreCase(FlagEncoderFactory.BIKE2)) {
						mBike2Encoding = true;
					}
					break;
                case FlagEncoderFactory.MOUNTAINBIKE:
                    addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_MTB, types);
                    break;
                case FlagEncoderFactory.RACINGBIKE:
                    addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_RACING, types);
                    break;
                case FlagEncoderFactory.FOOT:
                    addTrackType(GeoDataExtra.VALUE_RTE_TYPE_FOOT_01, types);
                    break;
                default:
                    Logger.logW(TAG, "getTrackTypes()," +
                            "unsupported type:" + encType);
                    break;
            }
        }

        // convert to array and return
        int[] typesA = new int[types.size()];
        for (int i = 0, m = types.size(); i < m; i++) {
            typesA[i] = types.get(i);
        }
        return typesA;
    }

	/**
	 * Add certain type into possible routing types.
	 * @param type type to add
	 * @param types types container
	 */
	private void addTrackType(int type, List<Integer> types) {
		if (!types.contains(type)) {
			types.add(type);
		}
	}

    @Override
    public Intent getIntentForSettings() {
        Intent intent = new Intent();
        intent.setClass(this, SettingsActivity.class);
        return intent;
    }

    @Override
    public int getNumOfTransitPoints() {
        return 2;
    }

    @Override
    public Track computeTrack(LocusUtils.LocusVersion lv, ComputeTrackParameters params) {
        if (getGraphHooper() == null) {
            Logger.logW(TAG, "computeTrack(" + params + ")," +
                    "unable to initialize GraphHopper instance");
            return null;
        }

        // define vehicle
        String vehicle = FlagEncoderFactory.CAR;
		String weighting = "";
		switch (params.getType()) {
			case GeoDataExtra.VALUE_RTE_TYPE_CAR:
				vehicle = FlagEncoderFactory.CAR;
				break;
			case GeoDataExtra.VALUE_RTE_TYPE_CAR_FAST:
				vehicle = FlagEncoderFactory.CAR;
				weighting = "fastest";
				break;
			case GeoDataExtra.VALUE_RTE_TYPE_CAR_SHORT:
				vehicle = FlagEncoderFactory.CAR;
				weighting = "shortest";
				break;
            case GeoDataExtra.VALUE_RTE_TYPE_MOTORCYCLE:
                vehicle = FlagEncoderFactory.MOTORCYCLE;
                break;
            case GeoDataExtra.VALUE_RTE_TYPE_CYCLE:
            	if (mBike2Encoding) {
					vehicle = FlagEncoderFactory.BIKE2;
				} else {
					vehicle = FlagEncoderFactory.BIKE;
				}
				break;
            case GeoDataExtra.VALUE_RTE_TYPE_CYCLE_FAST:
				if (mBike2Encoding) {
					vehicle = FlagEncoderFactory.BIKE2;
				} else {
					vehicle = FlagEncoderFactory.BIKE;
				}
				weighting = "fastest";
				break;
			case GeoDataExtra.VALUE_RTE_TYPE_CYCLE_SHORT:
				if (mBike2Encoding) {
					vehicle = FlagEncoderFactory.BIKE2;
				} else {
					vehicle = FlagEncoderFactory.BIKE;
				}
				weighting = "shortest";
				break;
            case GeoDataExtra.VALUE_RTE_TYPE_CYCLE_MTB:
                vehicle = FlagEncoderFactory.MOUNTAINBIKE;
                break;
            case GeoDataExtra.VALUE_RTE_TYPE_CYCLE_RACING:
                vehicle = FlagEncoderFactory.RACINGBIKE;
                break;
            case GeoDataExtra.VALUE_RTE_TYPE_FOOT_01:
				vehicle = FlagEncoderFactory.FOOT;
				break;
		}

        // finally compute
        return calcPath(params.getLocations(),
                vehicle, weighting,
                params.hasDirection() ? params.getCurrentDirection() : Double.NaN,
                params.isComputeInstructions());
    }

    /**
     * Lazy loader from compute engine
     * @return instance of GraphHooper service
     */
    private GraphHopper getGraphHooper() {
		// check current selected map
		File routingItem = Utils.getCurrentRoutingItem(this);
		if (routingItem == null) {
			throw new IllegalArgumentException("No valid routing item selected");
		}

		// initialize GraphHooper if required
		try {
			if (mHopper == null || mLastRoutingItem == null || !routingItem.equals(mLastRoutingItem)) {
				// reset parameters
				mBike2Encoding = false;

				// initialize graphHopper
				GraphHopper gh = new GraphHopper().forMobile();
				setGraphHopperProperties(gh, routingItem);
				boolean load = gh.load(routingItem.getAbsolutePath());
				Logger.logD(TAG, "found graph " + gh.getGraphHopperStorage() + ", " +
						"nodes:" + gh.getGraphHopperStorage().getNodes() + ", " +
						"path:" + routingItem.getAbsolutePath() + ", " +
						"load:" + load);

				// store result
				if (load) {
					mHopper = gh;
					mLastRoutingItem = routingItem;
				} else {
					mHopper = null;
				}
			}
		} catch (IOException e) {
			Logger.logE(TAG, "getGraphHooper()", e);
			mHopper = null;
		}

        // return initialized GraphHopper
        return mHopper;
    }

	/**
	 * Set parameters based on properties file.
	 * @param gh graphHopper instance
	 * @throws IOException excetion in case of any problems with loading properties file
	 */
	private void setGraphHopperProperties(GraphHopper gh, File routingItem) throws IOException {
		// set default parameters
		gh.setEnableInstructions(true);
		gh.setAllowWrites(false);
		gh.setCHEnabled(false);
		gh.setElevation(false);

		// load properties file
		byte[] filePropData = readFile(new File(routingItem, "properties"));
		if (filePropData == null || filePropData.length == 0) {
			return;
		}
		String fileProp = new String(filePropData);

		// test weighting
		Pattern patternWei = Pattern.compile("graph\\.ch\\.weightings=\\[(.*)\\]");
		Matcher matcherWei = patternWei.matcher(fileProp);
		if (matcherWei.find()) {
			gh.setCHEnabled(matcherWei.group(1).length() > 0);
		}

		// test elevation
		Pattern patternEle = Pattern.compile("graph\\.dimension=(\\d)");
		Matcher matcherEle = patternEle.matcher(fileProp);
		if (matcherEle.find()) {
			gh.setElevation(Integer.parseInt(matcherEle.group(1)) > 2);
		}
	}

	/**
	 * Read content of a file into in-memory byte array.
	 * @param file file to load
	 * @return loaded data
	 * @throws IOException exception in case of any problem
	 */
	private byte[] readFile(File file) throws IOException {
		// Open file
		RandomAccessFile f = null;
		try {
			// Get and check length
			f = new RandomAccessFile(file, "r");
			long longlength = f.length();
			int length = (int) longlength;
			if (length != longlength)
				throw new IOException("File size >= 2 GB");
			// Read file and return data
			byte[] data = new byte[length];
			f.readFully(data);
			return data;
		} finally {
			if (f != null) {
				f.close();
			}
		}
	}

    private Track calcPath(Location[] locs, String vehicle, String weighting,
            double firstPointDirection, boolean instructions) {
        Logger.logD(TAG, "calculating path for " + vehicle);
        StopWatch sw = new StopWatch().start();

        // define request and it's parameters
        GHRequest req = new GHRequest(2);
        for (int i = 0, m = locs.length; i < m; i++) {
            Location loc = locs[i];
            if (i == 0) {
                req.addPoint(new GHPoint(loc.getLatitude(), loc.getLongitude()), firstPointDirection);
            } else {
                req.addPoint(new GHPoint(loc.getLatitude(), loc.getLongitude()));
            }
        }
        req.setAlgorithm(Parameters.Algorithms.DIJKSTRA_BI);
        req.setVehicle(vehicle);
        req.setWeighting(weighting);
        req.getHints().
                put("instructions", instructions).
                put("douglas.minprecision", 1);

        // execute request
        GHResponse resp = mHopper.route(req);
        float time = sw.stop().getSeconds();

        if (!resp.hasErrors()) {
			PathWrapper path = resp.getBest();
            Logger.logD(TAG, "found path with distance:" + path.getDistance()
                    / 1000f + ", nodes:" + path.getPoints().getSize() + ", time:"
                    + time + " " + resp.getDebugInfo());
            Logger.logD(TAG, "the route is " + (int) (path.getDistance() / 100) / 10f
                    + "km long, time:" + path.getTime() / 60000f + "min, debug:" + time);
            return createTrackFromResult(resp, instructions);
        } else {
			// notify on problems
			List<Throwable> errors = resp.getErrors();
			for (int i = 0, m = errors.size(); i < m; i++) {
				final Throwable error = errors.get(i);
				if (error instanceof IllegalArgumentException &&
						error.getMessage().contains("Cannot find point ")) {
					Handler handler = new Handler(Looper.getMainLooper());
					handler.post(new Runnable() {

						@Override
						public void run() {
							Toast.makeText(RoutingService.this, error.getMessage(), Toast.LENGTH_LONG).show();
						}
					});
				}
			}

			// print result and return
            Logger.logW(TAG, "Error:" + resp.getErrors());
            return null;
        }
    }

    private Track createTrackFromResult(GHResponse resp, boolean instructions) {
        // create new track
        Track track = new Track();
		PathWrapper path = resp.getBest();

        // add points by instructions if exists
        if (instructions) {
            int instrCount = path.getInstructions().size();
            ViaInstruction viaInstruction = null;
            for (int i = 0; i < instrCount; i++) {
                // get parameters
                Instruction inst = path.getInstructions().get(i);
                PointList points = inst.getPoints();
                Location firstLoc = createLocation(points, 0);

                // store "via instruction" on next cycle
                if (inst instanceof ViaInstruction) {
                    viaInstruction = (ViaInstruction) inst;
                    continue;
                }

                // create instruction waypoint
                Waypoint wpt = new Waypoint();
                wpt.setLocation(firstLoc);
                wpt.addParameter(GeoDataExtra.PAR_RTE_POINT_ACTION,
                        graphHopperActionToLocus(inst).getId());

                // set name
                if (inst.getName().length() > 0) {
                    wpt.addParameter(GeoDataExtra.PAR_RTE_STREET, inst.getName());
                }

                // set speed and distance parameters
                wpt.addParameter(GeoDataExtra.PAR_RTE_DISTANCE_F,
                        Float.toString((float) inst.getDistance()));
                wpt.addParameter(GeoDataExtra.PAR_RTE_TIME_I,
                        (int) (inst.getTime() / 1000.0));
                track.getWaypoints().add(wpt);

                // use "Via instruction"
                if (viaInstruction != null) {
                    wpt.addParameter(GeoDataExtra.PAR_RTE_POINT_ACTION,
							PointRteAction.PASS_PLACE.getId());
                    viaInstruction = null;
                }

                // add points from instruction. Add also first point, so waypoint can be correctly indexed
                // and attached to correct point on track
                addPointsToTrack(track, inst.getPoints());
            }
        }

        // check if track already contain any data. If so, they're from instructions,
        // so we don't need to continue
        if (track.getPointsCount() > 0) {
            return track;
        }

        // add all points
        addPointsToTrack(track, path.getPoints());

        // return generated track
        return track;
    }

    private void addPointsToTrack(Track track, PointList points) {
        for (int i = 0, m = points.size(); i < m; i++) {
            track.getPoints().add(createLocation(points, i));
        }
    }

    /**
     * Generate simple location object from coordinates.
     * @param points list of points
     * @param index index of wanted point
     * @return generated location
     */
    private Location createLocation(PointList points, int index) {
        Location loc = new Location();
        loc.setLatitude(points.getLatitude(index));
        loc.setLongitude(points.getLongitude(index));
        return loc;
    }

    /**
     * Convert GraphHopper action to Locus action.
     * @param inst instruction on certain place
     * @return code of action in Locus system
     */
    private PointRteAction graphHopperActionToLocus(Instruction inst) {
        // get roundabout
        if (inst.getSign() == Instruction.USE_ROUNDABOUT) {

            // handle special instruction. If roundabout is not defined by it, rather ignore it,
            // then notify incorrect exit
            if (inst instanceof RoundaboutInstruction) {
                RoundaboutInstruction instRb = (RoundaboutInstruction) inst;
                return PointRteAction.getActionRoundabout(instRb.getExitNumber());
            } else {
                Logger.logW(TAG, "graphHopperActionToLocus(" + inst + "), " +
                        "invalid Roundabout instruction");
                return PointRteAction.NO_MANEUVER;
            }
        }

        // handle common instructions
        switch(inst.getSign()) {

            // TURN RIGHT

            case Instruction.TURN_SLIGHT_RIGHT:
                return PointRteAction.RIGHT_SLIGHT;
            case Instruction.TURN_RIGHT:
                return PointRteAction.RIGHT;
            case Instruction.TURN_SHARP_RIGHT:
                return PointRteAction.RIGHT_SHARP;

            // TURN LEFT

            case Instruction.TURN_SLIGHT_LEFT:
                return PointRteAction.LEFT_SLIGHT;
            case Instruction.TURN_LEFT:
                return PointRteAction.LEFT;
            case Instruction.TURN_SHARP_LEFT:
                return PointRteAction.LEFT_SHARP;

            // VARIOUS

            case Instruction.CONTINUE_ON_STREET:
                return PointRteAction.CONTINUE_STRAIGHT;
			case Instruction.KEEP_RIGHT:
				return PointRteAction.STAY_RIGHT;
			case Instruction.KEEP_LEFT:
				return PointRteAction.STAY_LEFT;

            case Instruction.FINISH:
                return PointRteAction.ARRIVE_DEST;
            case Instruction.REACHED_VIA:
                return PointRteAction.PASS_PLACE;

            // IGNORED

            case Instruction.LEAVE_ROUNDABOUT:
                return PointRteAction.NO_MANEUVER;
            default:
                return PointRteAction.NO_MANEUVER;
        }
    }
}

File: app/src/main/java/com/asamm/locus/addon/graphhopper/RoutingService.kt
/**
 * Created by menion on 9. 7. 2014.
 * Class is part of Locus project
 */
package com.asamm.locus.addon.graphhopper

import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.graphhopper.GHRequest
import com.graphhopper.GHResponse
import com.graphhopper.GraphHopper
import com.graphhopper.routing.util.DefaultFlagEncoderFactory
import com.graphhopper.routing.util.EncodingManager
import com.graphhopper.routing.util.FlagEncoderFactory
import com.graphhopper.routing.weighting.FastestWeighting
import com.graphhopper.routing.weighting.ShortestWeighting
import com.graphhopper.storage.RAMDirectory
import com.graphhopper.storage.StorableProperties
import com.graphhopper.util.*
import com.graphhopper.util.shapes.GHPoint
import locus.api.android.features.computeTrack.ComputeTrackParameters
import locus.api.android.features.computeTrack.ComputeTrackService
import locus.api.android.objects.LocusVersion
import locus.api.objects.extra.GeoDataExtra
import locus.api.objects.extra.Location
import locus.api.objects.extra.PointRteAction
import locus.api.objects.geoData.Point
import locus.api.objects.geoData.Track
import locus.api.utils.Logger
import java.io.File
import java.io.IOException
import java.util.*

/**
 * Main routing service that do all heavy work with compute of required route.
 */
class RoutingService : ComputeTrackService() {

    // instance of GraphHooper engine
    private var hopper: GraphHopper? = null
        get() {
            val routingItem = Utils.getCurrentRoutingItem(this)?.absoluteFile
                    ?: throw IllegalArgumentException("No valid routing item selected")
            try {
                // check current selected map
                if (field == null || routingItem != lastRoutingItem) {
                    bike2Encoding = false

                    // initialize GraphHooper
                    val gh = GraphHopper().forMobile()

                    // reset parameters
                    setGraphHopperProperties(gh, routingItem)

                    // initialize graphHopper
                    val load = gh.load(routingItem.path)
                    Logger.logD(TAG, "found graph ${gh.graphHopperStorage}, " +
                            "nodes: ${gh.graphHopperStorage.nodes}, " +
                            "path: ${routingItem.path}, " +
                            "load: $load")

                    // store result
                    if (load) {
                        field = gh
                        lastRoutingItem = routingItem
                    } else {
                        field = null
                    }
                }
            } catch (e: IOException) {
                Logger.logE(TAG, "getGraphHooper()", e)
                field = null
            }

            // return initialized GraphHopper
            return field
        }

    // remember last used map path
    private var lastRoutingItem: File? = null

    // flag if we have "bike2" mode in latest loaded file
    private var bike2Encoding = false

    override val attribution: String
        get() = "Powered by <a href=\"https://graphhopper.com/#enterprise\">GraphHopper API</a>"

    override val trackTypes: IntArray
        get() {
            // initialize graphHopper instance
            val hopper = hopper
            if (hopper == null) {
                Logger.logW(TAG, "getTrackTypes(), unable to initialize GraphHopper instance")
                return IntArray(0)
            }

            // load encoders
            val types = ArrayList<Int>()
            val em = hopper.encodingManager
            val profiles = hopper.chFactoryDecorator.chProfiles

            // get all encoders
            for (enc in em.fetchEdgeEncoders()) {

                // add car
                when (val encType = enc.toString()) {
                    FlagEncoderFactory.BIKE,
                    FlagEncoderFactory.BIKE2 -> {
                        var bikeAdded = false
                        for (profile in profiles) {
                            val weighting = profile.weighting
                            if (weighting is FastestWeighting) {
                                addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_FAST, types)
                                bikeAdded = true
                            } else if (weighting is ShortestWeighting) {
                                addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_SHORT, types)
                                bikeAdded = true
                            }
                        }

                        // add basic "bike" if no bike type exists
                        if (!bikeAdded) {
                            addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE, types)
                        }

                        // enable "Bike2" encoding
                        if (encType.equals(FlagEncoderFactory.BIKE2, ignoreCase = true)) {
                            bike2Encoding = true
                        }
                    }
                    FlagEncoderFactory.CAR -> {
                        var carAdded = false
                        for (profile in profiles) {
                            val weighting = profile.weighting
                            if (weighting is FastestWeighting) {
                                addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CAR_FAST, types)
                                carAdded = true
                            } else if (weighting is ShortestWeighting) {
                                addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CAR_SHORT, types)
                                carAdded = true
                            }
                        }

                        // add basic "car" if no car type exists
                        if (!carAdded) {
                            addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CAR, types)
                        }
                    }
                    FlagEncoderFactory.FOOT -> {
                        addTrackType(GeoDataExtra.VALUE_RTE_TYPE_FOOT_01, types)
                    }
                    FlagEncoderFactory.HIKE -> {
                        addTrackType(GeoDataExtra.VALUE_RTE_TYPE_FOOT_02, types)
                    }
                    FlagEncoderFactory.MOTORCYCLE -> {
                        addTrackType(GeoDataExtra.VALUE_RTE_TYPE_MOTORCYCLE, types)
                    }
                    FlagEncoderFactory.MOUNTAINBIKE -> {
                        addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_MTB, types)
                    }
                    FlagEncoderFactory.RACINGBIKE -> {
                        addTrackType(GeoDataExtra.VALUE_RTE_TYPE_CYCLE_RACING, types)
                    }
                    else -> Logger.logW(TAG, "getTrackTypes(), unsupported type:$encType")
                }
            }

            // convert to array and return
            return types.toIntArray()
        }

    /**
     * Add certain type into possible routing types.
     *
     * @param type  type to add
     * @param types types container
     */
    private fun addTrackType(type: Int, types: MutableList<Int>) {
        if (!types.contains(type)) {
            types.add(type)
        }
    }

    override val intentForSettings: Intent
        get() = Intent().apply {
            setClass(this@RoutingService, SettingsActivity::class.java)
        }

    override fun computeTrack(lv: LocusVersion?, params: ComputeTrackParameters): Track? {
        // check instance of GraphHopper
        if (hopper == null) {
            Logger.logW(TAG, "computeTrack($params), unable to initialize GraphHopper instance")
            return null
        }

        // define vehicle
        var vehicle = FlagEncoderFactory.CAR
        var weighting = ""
        when (params.type) {
            GeoDataExtra.VALUE_RTE_TYPE_CAR -> {
                vehicle = FlagEncoderFactory.CAR
            }
            GeoDataExtra.VALUE_RTE_TYPE_CAR_FAST -> {
                vehicle = FlagEncoderFactory.CAR
                weighting = "fastest"
            }
            GeoDataExtra.VALUE_RTE_TYPE_CAR_SHORT -> {
                vehicle = FlagEncoderFactory.CAR
                weighting = "shortest"
            }
            GeoDataExtra.VALUE_RTE_TYPE_CYCLE -> {
                vehicle = if (bike2Encoding) {
                    FlagEncoderFactory.BIKE2
                } else {
                    FlagEncoderFactory.BIKE
                }
            }
            GeoDataExtra.VALUE_RTE_TYPE_CYCLE_FAST -> {
                vehicle = if (bike2Encoding) {
                    FlagEncoderFactory.BIKE2
                } else {
                    FlagEncoderFactory.BIKE
                }
                weighting = "fastest"
            }
            GeoDataExtra.VALUE_RTE_TYPE_CYCLE_SHORT -> {
                vehicle = if (bike2Encoding) {
                    FlagEncoderFactory.BIKE2
                } else {
                    FlagEncoderFactory.BIKE
                }
                weighting = "shortest"
            }
            GeoDataExtra.VALUE_RTE_TYPE_CYCLE_MTB -> {
                vehicle = FlagEncoderFactory.MOUNTAINBIKE
            }
            GeoDataExtra.VALUE_RTE_TYPE_CYCLE_RACING -> {
                vehicle = FlagEncoderFactory.RACINGBIKE
            }
            GeoDataExtra.VALUE_RTE_TYPE_FOOT_01 -> {
                vehicle = FlagEncoderFactory.FOOT
            }
            GeoDataExtra.VALUE_RTE_TYPE_FOOT_02 -> {
                vehicle = FlagEncoderFactory.HIKE
            }
            GeoDataExtra.VALUE_RTE_TYPE_MOTORCYCLE -> {
                vehicle = FlagEncoderFactory.MOTORCYCLE
            }
        }

        // finally compute
        return calcPath(params.locations,
                vehicle, weighting,
                if (params.hasDirection) params.currentDirection.toDouble() else java.lang.Double.NaN,
                params.isComputeInstructions)
    }

    private fun calcPath(locs: Array<Location>, vehicle: String, weighting: String,
            firstPointDirection: Double, instructions: Boolean): Track? {
        Logger.logD(TAG, "calculating path for $vehicle")
        val sw = StopWatch().start()

        // define request and it's parameters
        val req = GHRequest(2)
        for (i in locs.indices) {
            val loc = locs[i]
            if (i == 0) {
                req.addPoint(GHPoint(loc.latitude, loc.longitude), firstPointDirection)
            } else {
                req.addPoint(GHPoint(loc.latitude, loc.longitude))
            }
        }

        req.vehicle = vehicle
        req.weighting = weighting
        req.hints.put("instructions", instructions).put("douglas.minprecision", 1)

        // execute request
        val resp = hopper!!.route(req)
        val time = sw.stop().seconds

        if (!resp.hasErrors()) {
            val path = resp.best
            Logger.logD(TAG, "found path with " +
                    "distance:" + path.distance / 1000f + ", " +
                    "nodes: ${path.points.size}, " +
                    "time: $time, ${resp.debugInfo}")
            Logger.logD(TAG, "the route " +
                    "distance: ${(path.distance / 100).toInt() / 10f} km, " +
                    "time: ${path.time / 60000f} min, " +
                    "debug: $time")
            return createTrackFromResult(resp, instructions)
        } else {
            // notify on problems
            val errors = resp.errors
            for (i in 0 until errors.size) {
                val error = errors[i]
                if (error is IllegalArgumentException && error.message?.contains("Cannot find point ") == true) {
                    val handler = Handler(Looper.getMainLooper())
                    handler.post { Toast.makeText(this@RoutingService, error.message, Toast.LENGTH_LONG).show() }
                }
            }

            // print result and return
            Logger.logW(TAG, "Error:" + resp.errors)
            return null
        }
    }

    private fun createTrackFromResult(resp: GHResponse, instructions: Boolean): Track {
        // create new track
        val track = Track()
        val path = resp.best

        // add points by instructions if exists
        if (instructions) {
            val instrCount = path.instructions.size
            var viaInstruction: ViaInstruction? = null
            for (i in 0 until instrCount) {
                // get parameters
                val inst = path.instructions[i]
                val points = inst.points
                val firstLoc = createLocation(points, 0)

                // store "via instruction" on next cycle
                if (inst is ViaInstruction) {
                    viaInstruction = inst
                    continue
                }

                // create instruction waypoint
                val pt = Point()
                pt.location = firstLoc
                pt.addParameter(GeoDataExtra.PAR_RTE_POINT_ACTION,
                        graphHopperActionToLocus(inst).id)

                // set name
                if (inst.name.isNotEmpty()) {
                    pt.addParameter(GeoDataExtra.PAR_RTE_STREET, inst.name)
                }

                // set speed and distance parameters
                pt.addParameter(GeoDataExtra.PAR_RTE_DISTANCE_F,
                        inst.distance.toString())
                pt.addParameter(GeoDataExtra.PAR_RTE_TIME_I,
                        (inst.time / 1000.0).toInt())
                track.waypoints.add(pt)

                // use "Via instruction"
                if (viaInstruction != null) {
                    pt.addParameter(GeoDataExtra.PAR_RTE_POINT_ACTION,
                            PointRteAction.PASS_PLACE.id)
                    viaInstruction = null
                }

                // add points from instruction. Add also first point, so waypoint can be correctly indexed
                // and attached to correct point on track
                addPointsToTrack(track, inst.points)
            }
        }

        // check if track already contain any data. If so, they're from instructions,
        // so we don't need to continue
        if (track.pointsCount > 0) {
            return track
        }

        // add all points
        addPointsToTrack(track, path.points)

        // return generated track
        return track
    }

    /**
     * Add pack of generated points into existing track.
     *
     * @param track track object
     * @param points loaded points
     */
    private fun addPointsToTrack(track: Track, points: PointList) {
        for (i in 0 until points.size) {
            track.points.add(createLocation(points, i))
        }
    }

    /**
     * Generate simple location object from coordinates.
     *
     * @param points list of points
     * @param index  index of wanted point
     * @return generated location
     */
    private fun createLocation(points: PointList, index: Int): Location {
        return Location().apply {
            longitude = points.getLongitude(index)
            latitude = points.getLatitude(index)
        }
    }

    /**
     * Convert GraphHopper action to Locus action.
     *
     * @param inst instruction on certain place
     * @return code of action in Locus system
     */
    private fun graphHopperActionToLocus(inst: Instruction): PointRteAction {
        // get roundabout
        if (inst.sign == Instruction.USE_ROUNDABOUT) {

            // handle special instruction. If roundabout is not defined by it, rather ignore it,
            // then notify incorrect exit
            return if (inst is RoundaboutInstruction) {
                PointRteAction.getActionRoundabout(inst.exitNumber)
            } else {
                Logger.logW(TAG, "graphHopperActionToLocus($inst), invalid Roundabout instruction")
                PointRteAction.NO_MANEUVER
            }
        }

        // handle common instructions
        when (inst.sign) {

            // TURN RIGHT

            Instruction.TURN_SLIGHT_RIGHT -> return PointRteAction.RIGHT_SLIGHT
            Instruction.TURN_RIGHT -> return PointRteAction.RIGHT
            Instruction.TURN_SHARP_RIGHT -> return PointRteAction.RIGHT_SHARP

            // TURN LEFT

            Instruction.TURN_SLIGHT_LEFT -> return PointRteAction.LEFT_SLIGHT
            Instruction.TURN_LEFT -> return PointRteAction.LEFT
            Instruction.TURN_SHARP_LEFT -> return PointRteAction.LEFT_SHARP

            // VARIOUS

            Instruction.CONTINUE_ON_STREET -> return PointRteAction.CONTINUE_STRAIGHT
            Instruction.KEEP_RIGHT -> return PointRteAction.STAY_RIGHT
            Instruction.KEEP_LEFT -> return PointRteAction.STAY_LEFT

            Instruction.FINISH -> return PointRteAction.ARRIVE_DEST
            Instruction.REACHED_VIA -> return PointRteAction.PASS_PLACE

            // IGNORED

            Instruction.LEAVE_ROUNDABOUT -> return PointRteAction.NO_MANEUVER
            else -> return PointRteAction.NO_MANEUVER
        }
    }

    //*************************************************
    // ENGINE
    //*************************************************

    /**
     * Set parameters based on properties file.
     *
     * @param gh graphHopper instance
     * @throws IOException exception in case of any problems with loading properties file
     */
    private fun setGraphHopperProperties(gh: GraphHopper, routingItem: File) {
        // set default parameters
        gh.isAllowWrites = false

        // load properties file
        val dir = RAMDirectory(routingItem.path, true)
        val properties = StorableProperties(dir)
        if (!properties.loadExisting()) {
            throw IllegalStateException("Cannot load properties to fetch EncodingManager configuration at: ${dir.location}")
        }

        // following code is not necessary - GraphHopper will create proper EncodingManager on demand,
        // but as in any case we load properties to configure CH and elevation, let's set EncodingManager explicitly
        val builder = EncodingManager.Builder()
        // enableInstructions is true by default, but to make sure
        builder.setEnableInstructions(true)
        properties.get("graph.encoded_values").takeIf { it.isNotBlank() }?.let {
            builder.addAll(gh.encodedValueFactory, it)
        }
        // GraphHopper doesn't expose flagEncoderFactory, but it is ok - stateless DefaultFlagEncoderFactory is used.
        properties.get("graph.flag_encoders").takeIf { it.isNotBlank() }?.let {
            builder.addAll(DefaultFlagEncoderFactory(), it)
        }

        gh.encodingManager = builder.build()

        gh.isCHEnabled = (properties.get("prepare.ch.done") == "true")
        gh.setElevation(properties.get("prepare.elevation_interpolation.done") == "true")
    }

    companion object {

        // tag for logger
        private const val TAG = "RoutingService"
    }
}
